src/Simulation/Simulators/OpenSystemsSimulator/DataModel/ComplexArrayConverter.cs

```diff
@@ -0,0 +1,250 @@
+        {
+            var element = data[idx];
+            array[idx, 0] = element.Real;
+            array[idx, 1] = element.Imaginary;
+        }
+        return np.reshape(array, dims.Concat(new [] { 2 }).ToArray());
+    }
+
+    public override void Write(Utf8JsonWriter writer, NDArray value, JsonSerializerOptions options)
+    {
+        // Before proceeding, check that `value` is complex-like. That is,
+        // that `value` is of dtype double, and has a trailing dimension
+        // of length 2.
+        if (!value.IsComplexLike())
+        {
+            throw new ArgumentException($"Cannot serialize ndarray, as it is not complex-like: {value}");
+        }
+
+        writer.WriteStartObject();
+            writer.WriteNumber("v", 1);
+
+            writer.WritePropertyName("dim");
+            writer.WriteStartArray();
+                foreach (var dim in value.shape[0..^1])
+                {
+                    writer.WriteNumberValue(dim);
+                }
+            writer.WriteEndArray();
+
+            writer.WritePropertyName("data");
+            writer.WriteStartArray();
+                // By default, NumSharp reshapes in C-order, matching
+                // ndarray's logical ordering. Thus, we reshape down to
+                // a two-axis array, and loop over the left most axis
+                // (corresponding to elements of the serialized array),
+                // leaving the second axis (corresponding to
+                // real-vs-imag).
+                var nElements = value.shape[0..^1].Aggregate((acc, dim) => acc * dim);
+                var flattened = value.reshape((nElements, 2));
+                foreach (var idx in Enumerable.Range(0, flattened.shape[0]))
+                {
+                    var element = flattened[idx];
+                    // Each element is a JSON array `[real, imag]`.
+                    writer.WriteStartArray();
+                        writer.WriteNumberValue((double) element[0]);
+                        writer.WriteNumberValue((double) element[1]);
+                    writer.WriteEndArray();
+                }
+            writer.WriteEndArray();
+        writer.WriteEndObject();
+    }
+
+    public static (int nQubits, string kind, NDArray data) ReadQubitSizedArray(ref Utf8JsonReader reader, JsonSerializerOptions options)
+    {
+        var arrayConverter = new ComplexArrayConverter();
+        int? nQubits = null;
+        NDArray? data = null;
+        string? kind = null;
+
+        // We require that the reader be in the right state to read an
+        // object.
+        if (reader.TokenType != JsonTokenType.StartObject)
+        {
+            throw new JsonException();
+        }
+
+        while (reader.Read())
+        {
+            if (reader.TokenType == JsonTokenType.EndObject)
+            {
+                // We're at the end of the array, and can break out of the
+                // read loop.
+                break;
+            }
```

Comment:
(Readability nit) In a number of places, consider shortening to
```suggestion
        while (reader.Read() && (reader.TokenType != JsonTokenType.EndObject))
        {
```